### ISSUE TITLE
[managedcluster] defaulting webhooks

### ISSUE THREAD

**Issue by alexeldeib** [kubernetes-sigs/cluster-api-provider-azure#615]
/kind feature

**Describe the solution you'd like**
There are probably a fair number of properties we could default earlier in the pipeline. CNI, Load balancer SKU, networking fields, node SKU/default pool. Add a defaulting webhook and begin defaulting some of these fields.

**Comment by jroden** [comment]
Hi Ace!  I'm still in the process of getting a better understanding (of managedcluster, capi & capiz, and webhooks) before beginning managedcluster work

I've got a couple questions to help validate my understanding if you don't mind :smile: 
Would the flow of creating a managedcluster look like the following?
1. CAPI controller receives request to create a new managecluster with provider azure
2. CAPI controller (may) perform transformation and/or logic based on the request
3. CAPI controller sends request to CAPZ to create a new managedcluster
4. CAPZ processes request and creates new managedcluster

And by using a webhook we can default values between steps 3 & 4 rather than if the logic were embedded in the capz api itself (step 4)?  Is this what you were referring to when you mentioned `defaulting earlier in the pipeline`?

Thanks for the guidance! :smile:

**Comment by jroden** [comment]
So currently we are defaulting fields for managedCluster in the 'Process Item' step of the custom controller? And thus moving to webhook would default the fields even before the controller picks up the changes from the api server?  (snip taken from [here](https://github.com/kubernetes/sample-controller/blob/master/docs/images/client-go-controller-interaction.jpeg))
![image](https://user-images.githubusercontent.com/28738612/84599201-4c53ff00-ae3e-11ea-8b69-e38074793d95.png)

**Comment by alexeldeib** [comment]
> And the webhook would default values as part of step 1? As opposed to if the logic were baked into the CAPZ controller?

Correct.

> So currently we are defaulting fields for managedCluster in the 'Process Item' step of the custom controller? And thus moving to webhook would default the fields even before the controller picks up the changes from the api server?

Also correct.

There are a few cases here that are poorly modeled even though they are technically optional fields. A good example is networkPlugin. I don't think AKS API requires this field to be set, but the API will reject attempts to update this field. We need to model that in our CRD. On the validation side, we would want to prevent the ability to persist invalid configurations (like changing the networkPlugin on update). 

There are a fairly limited number of cases i'd enable defaulting for. We should probably default networkPlugin to Azure (I think having no network policy is valid). 

A good case for defaulting webhooks is the service CIDR and the DNS service IP for AKS. You have a good opportunity to remove this hack by taking the service CIDR and defaulting the DNS IP, + adding a validating webhook to enforce the correctness if the user provides both: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/aa979f27a021e4d87e2bf2926344833042f400c3/cloud/services/managedclusters/managedclusters.go#L146-L152

**Comment by jroden** [comment]
Thanks for the response!  This gives me a great direction to go :smile: 
I will provide an update once I have made some progress

**Comment by jroden** [comment]
Ok, so I've got the defaulting webhook & test setup.  Just need to add defaulting logic and then the validating webhook.  Any other cases for defaulting or validation that come to mind?  I will do some research from my end as well to determine if there are other specs that would make sense to include in the defaulting/validation webhooks.

Thanks for the help! :smile:

**Comment by jroden** [comment]
Hi Ace, Hope you're having a good weekend!  I've made a bit of progress and am hoping to get your thoughts on an approach I'm trying for defaulting the dnsIP:

I've updated the CRD for ManagedCluster to include DNSServiceIP in the spec.  Inside the ManagedCluster webhooks, I will obtain the serviceCidr from the associated CAPI cluster spec (still working through the specifics of how this would look).  Once I have access to both specs in the webhooks I can perform the defaulting/validating logic.

Would love to hear your thoughts on including DNSServiceIP in the ManagedCluster CRD and pulling the serviceCidr from the webhooks! :smile:

**Comment by jroden** [comment]
Or should we model the serviceCidr under the ManagedCluster Spec instead of pulling from CAPI Cluster Spec?  I'm not sure if referencing a separate object in a webhook is possible (Pulling serviceCidr from CAPI Cluster CR in the CAPZ ManagedCluster webhooks)

**Comment by alexeldeib** [comment]
I think we would probably need DNSServiceIP in AzureManagedControlPlane spec and validate against the CAPI Cluster. We should _probably_ avoid duplicating the service CIDR so we maintain a single source of truth.

> still working through the specifics of how this would look

If we do cross-object validation inside webhooks, we probably need something like this: 
https://github.com/alexeldeib/hook-test/blob/ca37300b9e10548a8cfbf24ba6f782469d92ec6c/webhook.go#L15
https://github.com/alexeldeib/hook-test/blob/ca37300b9e10548a8cfbf24ba6f782469d92ec6c/main.go#L68-L73

@CecileRobertMichon I recall we talked about cross-object validation in webhooks like this, do you think this use case is okay? I don't remember the takeaway.

**Comment by CecileRobertMichon** [comment]
I think cross-object validation is fine, it's cross-object _defaulting_ that can be dangerous (ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/623)

**Comment by alexeldeib** [comment]
Got it. So maybe we can’t actually get rid of the IP CIDR hack in the defaulted case, but we can at least handle validation in the webhook if the DNS IP is provided at create time. 

We’d also want to validate against changing these fields on updates

**Comment by CecileRobertMichon** [comment]
If the field in the other object that the default is based on is immutable (and there's validation in place on that object's webhook to make sure it fails update if it is changed), then I think it's okay to default too. I'm more concerned about a case where we default something based on another object and that other object is subsequently modified (ie. the default is no longer accurate).

**Comment by jroden** [comment]
Currently debugging an issue in my dev cluster, if anyone has any pointers that come to mind those would be appreciated :smile:
I'm trying to test the webhook in kind using tilt.  When checking the logs I see my webhook registering at the path, but I'm missing the part where the builder associates the crd with the webhook path.  For example, I'm expecting the logs to look like the following:
```sh
I0629 00:56:14.679705       7 webhook.go:102] controller-runtime/builder "msg"="Registering a mutating webhook"  "GVK"={"Group":"infrastructure.cluster.x-k8s.io","Version":"v1alpha3","Kind":"AzureMachine"} "path"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-azuremachine"
I0629 00:56:14.679766       7 server.go:121] controller-runtime/webhook "msg"="registering webhook"  "path"="/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-azuremachine"
```
But instead I only have the second line:
```sh
I0629 00:56:14.681560       7 server.go:121] controller-runtime/webhook "msg"="registering webhook"  "path"="/mutate-exp-cluster-x-k8s-io-x-k8s-io-v1alpha3-azuremanagedcontrolplane"
```
Any pointers or direction is greatly appreciated, will continue digging in the meantime :smile:

**Comment by jroden** [comment]
@alexeldeib Hi Ace!  qq- I am wondering which design approach is preferred for determining which cluster is associated with the amcp object during webhook runtime.  I like the first option since it doesn't make any assumptions, but I could be missing something in that assessment.  what are your thoughts?

1. Add ```clusterRef``` field to the amcp spec (similar to controlPlaneRef in [CAPI ClusterSpec](https://godoc.org/sigs.k8s.io/cluster-api/api/v1alpha3#ClusterSpec))
2. If there is only one cluster in the same namespace as amcp take that one, if multiple skip defaulting dnsServiceIp
3. Infer the cluster name from the amcp name

Thanks for the help! really enjoying this one :smile:

**Comment by alexeldeib** [comment]
Sorry, missed replying to this --

Maybe the DNS IP should stay inside reconcile. The canonical way to get the corresponding cluster would be the owner reference., but that can't be set until after webhook runtime. We can validate against updates in the validating webhook after setting it during reconcile.

**Comment by nader-ziada** [comment]
already fixed in PR #862 

/close
